File: 8. Linked List/BasicLinkedList.py
class Node:

    def __init__(self, data):

        self.data = data
        self.next = None


class LinkedList:

    def __init__(self):

        self.head = None
        self.tail = None

    def PushBack(self, value):

        new_node = Node(value)

        if self.head is None:
            self.head = new_node
            self.tail = new_node

        else:
            self.tail.next = new_node
            self.tail = new_node

    def PopBack(self):

        if self.isEmpty():
            print("Cannot pop from an empty list")
            return

        if self.head == self.tail:

            self.head = None
            self.tail = None

        else:

            temp = self.head

            while temp.next != self.tail:
                temp = temp.next

            temp.next = None
            self.tail = temp

    def isEmpty(self):

        return self.head is None

    def clear(self):

        self.head = None
        self.tail = None

    def print(self):

        temp = self.head
        while temp is not None:
            print(temp.data, end=" ")
            temp = temp.next

        print()

my_list = LinkedList()
my_list.PushBack(1)
my_list.PushBack(2)
my_list.PushBack(3)

my_list.print()  

my_list.PopBack()
my_list.print()  

my_list.clear()
my_list.print()  

File: 8. Linked List/BasicLinkedList.cpp
#include <iostream>

struct Node
{
    int Value;
    Node *Next;
};

typedef Node *nodePtr;

void PrintList(nodePtr n)
{
    while (n != NULL)
    {
        std::cout << n->Value << std::endl;
        n = n->Next;
    }
}

void InsertNodeAtFront(nodePtr *head, int value)
{
    nodePtr newNode = new Node();

    newNode->Value = value;
    newNode->Next = *head;

    *head = newNode;
}

void InsertNodeAtTheLast(nodePtr *head, int value)
{
    nodePtr newNode = new Node();

    newNode->Value = value;
    newNode->Next = NULL;

    if (*head == NULL)
    {
        *head = newNode;
        return;
    }

    Node *last = *head;

    while (last->Next != NULL)
    {
        last = last->Next;
    }

    last->Next = newNode;
}

int main()
{
    nodePtr head = new Node();
    nodePtr second = new Node();
    nodePtr third = new Node();

    head->Value = 1;
    second->Value = 2;
    third->Value = 3;

    head->Next = second;
    second->Next = third;
    third->Next = NULL;

    InsertNodeAtFront(&head, -1);
    InsertNodeAtTheLast(&third, 4);

    PrintList(head);

    return 0;
}